packages/client/helpers/functional-test/run-tests.ts

```diff
@@ -8,49 +12,133 @@ const allProviders = new Set(Object.values(Providers))
-  const unknownProviders = providers.filter((provider) => !allProviders.has(provider))
-  if (unknownProviders.length > 0) {
-    console.error(`Unknown providers: ${unknownProviders.join(', ')}`)
-    process.exit(1)
+  if (args['--provider']) {
+    const providers = args['--provider'] as Providers[]
+    const unknownProviders = providers.filter((provider) => !allProviders.has(provider))
+    if (unknownProviders.length > 0) {
+      console.error(`Unknown providers: ${unknownProviders.join(', ')}`)
+      process.exit(1)
+    }
+    jestCli = jestCli.withEnv({ ONLY_TEST_PROVIDERS: providers.join(',') })
   }
-  jestCli = jestCli.withEnv({ ONLY_TEST_PROVIDERS: providers.join(',') })
-}
 
-const codeTestCli = jestCli.withArgs(['--testPathIgnorePatterns', 'typescript'])
+  if (args['--data-proxy']) {
+    if (!fs.existsSync(miniProxy.defaultServerConfig.cert)) {
+      await miniProxy.generateCertificates(miniProxy.defaultCertificatesConfig)
+    }
+
+    jestCli = jestCli.withEnv({
+      DATA_PROXY: 'true',
+    })
+
+    if (!args['--no-mini-proxy-default-ca']) {
+      jestCli = jestCli.withEnv({
+        NODE_EXTRA_CA_CERTS: miniProxy.defaultCertificatesConfig.caCert,
+      })
+    }
+
+    if (args['--edge-client']) {
+      jestCli = jestCli.withEnv({
+        TEST_DATA_PROXY_EDGE_CLIENT: 'true',
+      })
+    }
+
+    if (!args['--no-mini-proxy-server']) {
+      const qePath = await getBinaryForMiniProxy()
 
-try {
-  if (args['-u']) {
-    const snapshotUpdate = codeTestCli.withArgs(['-u']).withArgs(args['_'])
-    snapshotUpdate.withEnv({ UPDATE_SNAPSHOTS: 'inline' }).run()
-    snapshotUpdate.withEnv({ UPDATE_SNAPSHOTS: 'external' }).run()
-  } else {
-    if (!args['--types-only']) {
-      codeTestCli.withArgs(['--']).withArgs(args['_']).run()
+      miniProxyProcess = execa('mini-proxy', ['server', '-q', qePath], {
+        preferLocal: true,
+        stdio: 'inherit',
+        env: {
+          DEBUG: args['--mini-proxy-debug'] ? 'mini-proxy:*' : process.env.DEBUG,
+        },
+      })
     }
+  }
+
+  if (args['--edge-client'] && !args['--data-proxy']) {
+    throw new Error('--edge-client is only available when --data-proxy is used')
+  }
 
-    if (!args['--no-types']) {
-      jestCli.withArgs(['--', 'typescript']).run()
+  const codeTestCli = jestCli.withArgs(['--testPathIgnorePatterns', 'typescript'])
+
+  try {
+    if (args['-u']) {
+      const snapshotUpdate = codeTestCli.withArgs(['-u']).withArgs(args['_'])
+      snapshotUpdate.withEnv({ UPDATE_SNAPSHOTS: 'inline' }).run()
+      snapshotUpdate.withEnv({ UPDATE_SNAPSHOTS: 'external' }).run()
+    } else {
+      if (!args['--types-only']) {
+        codeTestCli.withArgs(['--']).withArgs(args['_']).run()
+      }
+
+      if (!args['--no-types']) {
+        jestCli.withArgs(['--', 'typescript']).run()
+      }
+    }
+  } catch (error) {
+    if (error.exitCode) {
+      // If it's execa error, exit without logging: we
```

Comment:
Would it be handy to log the error here just in case something gets swallowed without us knowing? I can imagine it causing some headaches.

Reply:
I thought so too when I first encountered this code, but in fact it wouldn't really be handy and would only introduce extra noise when any test fails without adding any new information.

Basically there are two possibilities:

1. The error is an actual error with useful information. In that case the check here will fail and the error will be re-thrown below and fully logged onto the screen.

2. The error is artificially raised by `execa` because of a non-zero exit code that the Jest process has terminated with. In that case, the error carries no useful information except the specific exit code. All of the context in the form of `stderr` output has already been printed onto the screen. The exit code doesn't really need to be printed (Jest wouldn't do this when running it without our wrapper, and it will be available in the shell regardless if you need to see it), but we have to make sure to exit with this code (which we do).

It would make sense to print it, though, if we, let's say, always exited with exit code 1 whenever Jest fails and not forwarded the code back to the shell.